news/models.py

```diff
@@ -17,33 +29,78 @@ class Entry(models.Model):
 
     """
 
+    class AlreadyApprovedError(Exception):
+        """The entry cannot be approved again."""
+
     slug = models.SlugField()
     title = models.CharField(max_length=255)
     description = models.TextField(blank=True, default="")
     author = models.ForeignKey(User, on_delete=models.CASCADE)
+    moderator = models.ForeignKey(
+        User,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="moderated_entries_set",
+    )
     external_url = models.URLField(blank=True, default="")
     image = models.ImageField(upload_to="news", null=True, blank=True)
     created_at = models.DateTimeField(default=now)
+    approved_at = models.DateTimeField(null=True, blank=True)
+    modified_at = models.DateTimeField(auto_now=True)
     publish_at = models.DateTimeField(default=now)
 
+    objects = EntryManager()
+
     class Meta:
         verbose_name_plural = "Entries"
 
     def __str__(self):
         return f"{self.title} by {self.author}"
 
+    @property
+    def is_approved(self):
+        return (
+            self.moderator is not None
+            and self.approved_at is not None
+            and self.approved_at <= now()
+        )
+
+    @property
+    def is_published(self):
+        return self.is_approved and self.publish_at <= now()
+
+    def approve(self, user):
+        """Mark this entry as approved by the given `user`."""
+        if self.is_approved:
+            raise self.AlreadyApprovedError()
+        self.moderator = user
+        self.approved_at = now()
+        self.save(update_fields=["moderator", "approved_at"])
+
     def save(self, *args, **kwargs):
         if not self.slug:
             self.slug = slugify(self.title)
         return super().save(*args, **kwargs)
 
-    @property
-    def published(self):
-        return self.publish_at <= now()
-
     def get_absolute_url(self):
         return reverse("news-detail", args=[self.slug])
 
+    def can_view(self, user):
+        return (
+            self.is_published
+            or user == self.author
```

Comment:
Should this include `user == self.moderator` as well? I'm not sure of the flow.

Reply:
I was imaging the "moderators" to be a Django group that we create manually in the admin, and this group would have granted all the individual perms for Entry, that is add, change, view and delete. So basically any user with the perm "view_entry" would be able to view it, and moderators would have this perm by getting the group granted.

I wouldn't use comparison against the original moderator user since that user could be long demoted from the role (for example, an ex-employee or similar).

Does that make sense?

Reply:
@nessita Makes sense to me! Thanks for clarifying.